Allow `DataParallel` and `jit.trace`

jatentaki opened the pull request:
1. Changed some tensors which used to be updated with in-place sliced mutation to instead concatenate lists of temporary results. This may have resulted in some increase of memory footprint, but the previous code was technically UB from the perspective of autograd (as far as I understand the documentation and pytorchers on slack), so now it plays nicely with `nn.DataParallel` and `jit.trace`.
2. Reorganized the tests, moved to top-level `tests` directory. The `all` runner script is ugly, but I couldn't get `python -m unittest discover` to work. Also note that `all` runs all tests in the directory, not all in the whole project
3. Lowered the memory footprint of batchnorm slightly by detaching a variable earlier.

Hopefully the commit messages are more descriptive of particular modifications. I tried to test this under `0.4.1` and `1.0 preview`, but please test this yourself before merging

jatentaki committed:
Reduce memory footprint by detaching field_norm early

jatentaki committed:
Move tests from convolution.py to conv_tests.py; modify the combination implementation to use cat instead of in-place array modification to allow use of jit

jatentaki committed:
Add test for wrapping SE3Convolutions in nn.DataParallel. Tests pass

jatentaki committed:
Modify batch normalization to use torch.cat instead of mutating tensors inplace

jatentaki committed:
Move tests to their own directory; rename tests

mariogeiger reviewed (changes requested):
Be careful, `field_mean` and `field_norm` are used later to normalize the fields.
We need the gradient to go trough them !

I think this modification is not good.

mariogeiger commented on se3cnn/kernel.py:
```diff
@@ -373,13 +375,19 @@ def combination(self, weight):
-                    kernel[si, sj] = ker.view_as(kernel[si, sj])
                 else:
-                    kernel[si, sj] = 0
+                    ker = torch.tensor(0.)
```
Please always use things like `weight.new_tensor` to ensure have the same `device` and `dtype`.
Then I think you cannot apply `view` on it. At least `torch.tensor(0.).view(2, 2)` crash on my computer.

mariogeiger replied:
You could add a test that has at least one `kij is None`.
`se3cnn.kernel.SE3Kernel([(1, 0)], [(1, 0), (1, 5)], 3)` does the trick

mariogeiger reviewed (changes requested):
Looks super cool. If jit works it's amazing.

Please test also the gradients. Please compare your implementation output+gradient with the current version. They must match.

mariogeiger commented:
I see that some code with `dyn_iso` is removed. Is it because you don't have the last commits from master ? I don't know if I have to worry about that or if it will be managed during merge.

jatentaki committed:
1) Skip DataParallel if multiple GPUs are not available
2) Add gradcheck for bnconv
3) Add a check for case when `kij` is None in `SE3Kernel.combine`. Currently lacks an assertion (only check if it doesn't throw)

jatentaki committed:
Fix issue with kij == None

jatentaki committed:
Implement test for SE3BatchNorm, give assertions more meaningful output in se3cnn/batchnorm.py

jatentaki committed:
1) Fix incorrect gradients in SE3BatchNorm
2) Incorrect gradients in SE3BNConv remain, they seem to be numerical issues. They also appear in the original implementation
3) Break some lines into multiple lines (especially trailing comment) to allow viewing on narrower screen

jatentaki commented:
I addressed most issues
1) The gradient indeed needs to flow there, fixed. Additionally, I added gradtests for both `SE3BatchNorm` and `SEBNConv`. The former passes, the latter fails due to what appears to numerical errors. Note that it also fails with the original implementation (current master). Do you have any idea how to address that?
2) Can you think of a better way to test correctness than to compare with the old version? It feels awkward to keep both for this purpose, plus realistically the older version is also not guaranteed to be bug-free. I would rather trust some actual sanity checks. Checking for equivariance follows this line of thought, but I agree more tests would be useful.
3) Fixed the `kij is None` issue. I did not use `new_tensor` but `torch.zeros(..., device=old.device, dtype=old.dtype)`, which effectively does the same thing.
4) I have no what this `dyn_iso` thing is, I have never touched it. I am trying to figure out how to pull your upstream changes to my fork.

jatentaki committed:
Fix bug where we tried to concatenate empty lists in eval mode

mariogeiger commented on tests/batchnorm_tests.py:
```diff
@@ -54,6 +54,19 @@ def unrotate(t):
+        x = torch.rand(1, sum(m * (2 * l + 1) for m, l in Rs_in), 6, 6, 6,
+                       requires_grad=True, dtype=torch.float64)
+
+        self.assertTrue(torch.autograd.gradcheck(conv, (x, ), eps=1))
```
It is not a linear function so it should not work with `eps=1`, did you try with a smaller value ?

mariogeiger commented on tests/batchnorm_tests.py:
```diff
@@ -54,6 +54,19 @@ def unrotate(t):
+        x = torch.rand(1, 4, 6, 6, 6,
+                       requires_grad=True, dtype=torch.float64)
+
+        self.assertTrue(torch.autograd.gradcheck(bn, (x, ), eps=1))
```
Same here, `y = x / sqrt(var(x))` is nonlinear therefore `eps` should be small

mariogeiger reviewed (commented):
If you want to check the derivative of BN you can also check with a very simple test vs hand calculus:
```python
x = torch.rand(3, requires_grad=True).view(1, 3, 1, 1, 1)
bn = se3cnn.SE3BatchNorm([(1, 3)], eps=0)
y = bn(x)

# by hand dyi/dxj = delta_ij / |x| - xi xj / |x|^3

jac_num = jacobian(y, x)  # github.com/mariogeiger/hessian
jac_cal = torch.eye(3) / x.norm() - x.view(-1, 1) * x.view(1, -1) / x.norm() ** 3

assert jac_cal == jac_num
```

For simplicity I took `eps=0` but one could also consider it into the calculus as well.

mariogeiger committed:
Update batchnorm_tests.py
